instance.go

```diff
@@ -0,0 +1,147 @@
+package pusherplatform
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+)
+
+var (
+	slashFoldingRegexp  regexp.Regexp
+	trailingSlashRegexp regexp.Regexp
+)
+
+func init() {
+	slashFoldingRegexp = *regexp.MustCompile("\\/+")
+	trailingSlashRegexp = *regexp.MustCompile("\\/$")
+}
+
+const (
+	hostBase = "pusherplatform.io"
+)
+
+// Instance allows making HTTP requests to a service
+type Instance interface {
+	Request(ctx context.Context, options RequestOptions) (*http.Response, error)
+	Authenticator() Authenticator
+}
+
+// InstanceOptions to initialize a new instance.
+type InstanceOptions struct {
+	// Instance locator unique to an app
+	Locator string
+	// Key unique to an app
+	Key string
+	// Service name to connect to
+	ServiceName string
+	// Version of service to connect to
+	ServiceVersion string
+
+	// Optional host
+	// If not provided, it will be constructed on the basis
+	// of the instance locator and the base host
+	Host string
+	// Optional BaseClient
+	// If not provided, it will be constructed
+	Client BaseClient
+}
+
+type instance struct {
+	instanceID      string
+	serviceName     string
+	serviceVersion  string
+	cluster         string
+	platformVersion string
+
+	keyID     string
+	keySecret string
+
+	authenticator Authenticator
+	client        BaseClient
+}
+
+// NewInstance creates a new instance satisfying the Instance interface
+func NewInstance(options InstanceOptions) (Instance, error) {
+	platformVersion, cluster, instanceID, err := getinstanceLocatorComponents(options.Locator)
+	if err != nil {
+		return nil, err
+	}
+
+	keyID, keySecret, err := getKeyComponents(options.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	var host string
+	if options.Host != "" {
+		host = options.Host
+	} else {
+		host = fmt.Sprintf("%s.%s", cluster, hostBase)
+	}
+
+	client := options.Client
+	if options.Client == nil {
+		client = NewBaseClient(BaseClientOptions{
+			Host: host,
+		})
+	}
+
+	return &instance{
+		instanceID:      instanceID,
+		serviceName:     options.ServiceName,
+		serviceVersion:  options.ServiceVersion,
+		cluster:         cluster,
+		platformVersion: platformVersion,
+		keyID:           keyID,
+		keySecret:       keySecret,
+		authenticator:   NewAuthenticator(instanceID, keyID, keySecret),
+		client:          client,
+	}, nil
+}
+
+// Request allows making HTTP requests to services
+func (i *instance) Request(ctx context.Context, options RequestOptions) (*http.Response, error) {
+	path := i.scopePath(options.Path)
+
+	var jwt *string
+	if options.Jwt == nil {
+		tokenResponse, err := i.Authenticator().GenerateAccessToken(AuthenticateOptions{Su: true})
```

Comment:
Should `Su: true` be hard-coded?

Reply:
Yeah, if there isn't a JWT, we have to fetch one.

Reply:
I'm not sure `Su: true` should be the "fallback" behaviour though. It's powerful enough that I think you should have to provide it as an explicit option or use a specialised function like `SuRequest` or whatever.

Reply:
Also, su tokens are (at least for now) a Chatkit-specific concern, so shouldn't _really_ be something that appears at all in a platform SDK. In the Chatkit Go SDK we should probably just have a `SuRequest` that sets up a `Request` call with the appropriate options.

Reply:
Hmmm, I'm quite sure I saw this in a platform library...